tests/server/test_pandas.py

```diff
@@ -0,0 +1,176 @@
+            geo_type=json["geo_type"],
+            geo_value=json["geo_value"],
+            direction=json["direction"],
+            issue=json["issue"],
+            lag=json["lag"],
+            value=json["value"],
+            stderr=json["stderr"],
+            sample_size=json["sample_size"],
+            missing_value=json["missing_value"],
+            missing_stderr=json["missing_stderr"],
+            missing_sample_size=json["missing_sample_size"],
+        )
+
+    @property
+    def signal_pair(self):
+        return f"{self.source}:{self.signal}"
+
+    @property
+    def geo_pair(self):
+        return f"{self.geo_type}:{self.geo_value}"
+
+    @property
+    def time_pair(self):
+        return f"{self.time_type}:{self.time_value}"
+
+    @property
+    def astuple(self):
+        return astuple(self)[1:]
+
+    @property
+    def aslist(self):
+        return list(self.astuple)
+
+
+class UnitTests(unittest.TestCase):
+    """Basic unit tests."""
+
+    def setUp(self):
+        """Perform per-test setup."""
+
+        # connect to the `epidata` database and clear the `covidcast` table
+        cnx = mysql.connector.connect(user="user", password="pass", host="delphi_database_epidata", database="epidata")
+        cur = cnx.cursor()
+        cur.execute("truncate table covidcast")
+        cur.execute('update covidcast_meta_cache set timestamp = 0, epidata = ""')
+        cnx.commit()
+        cur.close()
+
+        # make connection and cursor available to test cases
+        self.cnx = cnx
+        self.cur = cnx.cursor()
+
+    def tearDown(self):
+        """Perform per-test teardown."""
+        self.cur.close()
+        self.cnx.close()
+
+    def _insert_rows(self, rows: Iterable[CovidcastRow]):
+        sql = ",\n".join((str(r) for r in rows))
+        self.cur.execute(
+            f"""
+            INSERT INTO
+                `covidcast` (`id`, `source`, `signal`, `time_type`, `geo_type`,
+	            `time_value`, `geo_value`, `value_updated_timestamp`,
+                `value`, `stderr`, `sample_size`, `direction_updated_timestamp`,
+                `direction`, `issue`, `lag`, `is_latest_issue`, `is_wip`,`missing_value`,
+                `missing_stderr`,`missing_sample_size`)
+            VALUES
+            {sql}
+            """
+        )
+        self.cnx.commit()
+        return rows
+
+    def _rows_to_df(self, rows: Iterable[CovidcastRow]) -> pd.DataFrame:
+        columns = [
+            'id', 'source', 'signal', 'time_type', 'geo_type', 'time_value',
+            'geo_value', 'value_updated_timestamp', 'value', 'stderr',
+            'sample_size', 'direction_updated_timestamp', 'direction', 'issue',
+            'lag', 'is_latest_issue', 'is_wip', 'missing_value', 'missing_stderr',
+            'missing_sample_size'
+        ]
+        return pd.DataFrame.from_records([[i] + row.aslist for i, row in enumerate(rows, start=1)], columns=columns)
+
+    def test_as_pandas(self):
```

Comment:
given the name, I expected this test to call `as_pandas`, but it does not. on purpose?

Reply:
I agree it's weird that I just duplicated the innards of `as_pandas` here, but I couldn't think of a way to pass in my own `engine` variable into it.

Reply:
I modified this with some Flask testing magic, but now I'm running into an issue with SQLAlchemy 
```
    def as_pandas(query: str, params: Dict[str, Any], db_engine: Engine = engine, parse_dates: Optional[Dict[str, str]] = None, limit_rows = MAX_RESULTS+1) -> pd.DataFrame:
        try:
            query = limit_query(query, limit_rows)
            return pd.read_sql_query(text(str(query)), db_engine, params=params, parse_dates=parse_dates)
        except Exception as e:
>           raise DatabaseErrorException(str(e))
E           delphi.epidata.server._exceptions.DatabaseErrorException: 200 OK: database error: Execution failed on sql 'SELECT t.id, t.source, t.signal, t.time_type, t.geo_type, t.time_value, t.geo_value, t.value_updated_timestamp, t.value, t.stderr, t.sample_size, t.direction_updated_timestamp, t.direction, t.issue, t.lag, t.is_latest_issue, t.is_wip, t.missing_value, t.missing_stderr, t.missing_sample_size FROM covidcast t      LIMIT 10000001': Boolean value of this clause is not defined
```
`Boolean value of this clause is not defined` is pretty cryptic and the [documentation ](https://docs.sqlalchemy.org/en/13/core/tutorial.html#using-textual-sql) hasn't helped so far.

Reply:
Turned out the issue was that the `mysql.connector` used in the tests is different from the `sqlalchemy.engine` used in the actual server and `pd.read_sql_query` is only happy with the latter.